alphabase/io/tempmmap.py

```diff
@@ -65,12 +65,23 @@ def array(shape: tuple, dtype: np.dtype) -> np.ndarray:
         A tuple with the shape of the array.
     dtype : type
         The np.dtype of the array.
+    tmp_dir_name : str, optional
+        If specified the memory mapped array will be created in this directory. Defaults to None.
 
     Returns
     -------
     type
         A writable temporary mmapped array.
     """
+    # redefine the temporary directory if a new location is given otherwise read from global variable
```

Comment:
Yes. It had randint before, but I guess we modified it some time ago to reuse the folder. Now, using temp_dir_name, we can actually decide if we need to reuse the mmap.

Reply:
I think we should definetly aim for this!
I assumed the names of the mmap blobs themselve are already randomized?

Reply:
@GeorgWa ahh, good point.

Reply:
The folder names are still randomised. This is the default when producing a TemporaryDirectory using tempfile. 

![Screenshot 2024-05-08 at 10 53 11](https://github.com/MannLabs/alphabase/assets/15019107/ee2875c8-08f6-40ad-abec-9a26f6362fbe)

I am not sure why the file lock becomes an issue with multiple threads on linux. According to the documentation: The file is readable and writable only by the creating user ID. I was thinking that potentially the generated temp directories are only accessible by the thread that created them? Maybe someone more well-versed in Linux file permissions/thread operations would be able to give input on that.

Reply:
In general I think it would be good to find a robust solution that works across platforms and threads . I use the memory mapped temp arrays quite heavily in SPARCSpy and I guess I have a couple of unique use-cases.

1. Creating an empty memory mapped array that different processes can reconnect to (e.g. I have 4 channel images, each process modifies the information in one channel and then writes the results back to  the channel slot in the same memory mapped temp array)
2. individual processes have their own temp directories that are cleared as soon as they complete (this way we minimise disk usage as no longer needed things are cleared up straight away instead of at the end of the whole process)

The fixes with creating an empty memory mapped array and reconnecting to it (#151 and #152) addressed point 1. For point 2 I needed to be able to pass the specific path in which to produce a memory mapped temp array to ensure that it worked across threads on linux (#162). 

The implementation itself is great because you can use them just like numpy arrays but easily circumvent memory limitations. I have also found them to be fast and robust. So it would be great if we find a cleaner solution that maybe makes them more widely applicable.